moneymanagerex/moneymanagerex
# refactor(cmake,appveyor,travis): update build process and CI

AppVeyor changes:
* correctly initialize environment for different VisualStudio versions
* add support for different toolchains (120, 140, 141)
  and VS 2013/2015/2017 images
* add support for XP builds (toolchains like 140_xp...)
* add support for Debug builds
* build correct wxWidgets version for different toolchain/target/arch
* move from NMake to MSBuild to support parallel builds and toolchain
  selection
* more detailed package names
  (different toolchain builds share the same name!)
* auto detect and include in packages correct run-time DLLs if available
* include ucrt libraries for XP builds
* add checksums to the artifacts
* patch CMake FindwxWidgets module to support XP builds
- rename script to .appveyor.yml
- do not clone all wxWidgets history only selected version tag
- prefer x64 compiler to produce x86 builds (arch amd64_x86 in place x86)
- do not compress cache twice
- fix path order to fix git clone problems
- do not clone all MMEX versions only 10 last commits
- restore correct version wxWidgets from cache
- fix style of yml file

Travis changes:
* build packages for new distributions
* add priorities for build jobs
* set GitHub check result based on priority jobs only
* do not rename artifacts - use packages names generated by CMake
* add checksums to the artifacts
- do not upload Debug builds
- fix style of yml file

CMake scripts changes:
* do not build in-source - use only /build directory for all tasks
  (/bin removed too)
* add support for multi-configuration generators (remove hard-coded paths
  etc.) like VisualStudio
* move script parts related to specific sources into subdirectory scripts
* do not modify constants.h on-the-fly, use separate versions.h.in
* change version string (in toolchain only!)
  from 1.4.0-ALPHA-1 to 1.4.0~alpha1 to be Linux versionning compatible
* use static MMEXLib to compile sources once and link both mmex and mmex-tests
* use InstallRequiredSystemLibraries CMake module to detect and install
  correct versions of required run-time shared libraries on Windows
* Add settings summary at the end of the configuration phase (all messages
  moved there)
* replace /util/getDistroName.sh with official lsb_release tool incorporated
  into CMake script
* generate correct package names for different distributions with workaround
  for old CMake versions
* auto-detect correct package dependencies for Linux .deb/.rpm packages
* change po/en directory name to locale/en
- move main version numbers into project definition
- store git commit details in versions.h to not clutter command line
- add support for CMake 3.0 to support Debian jessie
- add new section for tuning settings when used from VS IDE
- set many new properties for Linux packages
- change variables prefix to MMEX_
- rename MMEXTest to mmex-tests
- fix style
- small fixes not listed here...

TODO:
- verify if CMake script can be auto loaded and used in the latest
  VisualStudio 2017 IDE
- add support for shared wxWindows (DLL) build for Windows
- replace cached wxWindows binaries with official distributed
  when available (me must compile for XP)
- fix mmex-tests (MMEX_BUILD_TEST=ON) builds (problems with CPPUnit)
- use correct .po/.mo filenames (do not reload message catalog for English
  but change language for the same catalog...)
- fix mmex_tests.rc


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/moneymanagerex/moneymanagerex/1302)
